sql/2021/resource-hints/pages_almanac_by_device_dns-prefetch_domains.sql

```diff
@@ -0,0 +1,67 @@
+#standardSQL
+# Most popular domains users dns-prefetch from
+
+# helper to create percent fields
+CREATE TEMP FUNCTION AS_PERCENT (freq FLOAT64, total FLOAT64) RETURNS FLOAT64 AS (
+  ROUND(SAFE_DIVIDE(freq, total), 4)
+);
+
+CREATE TEMPORARY FUNCTION getResourceHints(payload STRING, hint STRING)
+RETURNS STRING
+LANGUAGE js AS '''
+try {
+  var $ = JSON.parse(payload);
+  var almanac = JSON.parse($._almanac);
+  return JSON.stringify(almanac['link-nodes'].nodes.filter(link => link.rel.toLowerCase() == hint));
+} catch (e) {  
+  return "";
+}
+''' ;
+
+CREATE TEMPORARY FUNCTION getHrefs(payload STRING)
+RETURNS ARRAY<STRING>
+LANGUAGE js AS '''
+try {
+    var $ = JSON.parse(payload);
+    return $.map(n => n.href);
+} catch (e) {
+    return [];
+}
+''' ;
+
+SELECT
+  client,
+  domain,
+  COUNT(0) AS freq,
+  AS_PERCENT(COUNT(0), SUM(COUNT(0)) OVER()) AS pct
+FROM (
+    SELECT
+      client,
+      REGEXP_EXTRACT(href, r'^https?\:\/\/(.+?)\/') AS domain
+    FROM (
+        SELECT
+          client,
+          getHrefs(hints) AS value
+        FROM (
+            SELECT
+              _TABLE_SUFFIX AS client,
+              getResourceHints(payload, "dns-prefetch") AS hints
+            FROM
+              `httparchive.pages.2021_07_01_*`
+        )
+    )
+    CROSS JOIN UNNEST(value) AS href
+    WHERE
+      ARRAY_LENGTH(value) > 0
```

Comment:
Do we want to include `hints` that don't have hrefs in the final fractions? If yes, we would need `LEFT OUTER JOIN` without this `WHERE` condition.

```sql
SELECT id, COUNT(0)
FROM (
  WITH abc AS (
  SELECT 1 AS id, [1, 2, 3] AS vals
  UNION ALL
  SELECT 2 AS id, NULL AS vals
  )
  SELECT id, val
  FROM abc
  LEFT JOIN UNNEST(vals) AS val
  -- CROSS JOIN UNNEST(vals) AS val
)
GROUP BY id
```

Reply:
Thank you Nitin. We only want to include hints that have an `href` value as otherwise they would be invalid. However I am currently excluding relative hrefs (such as `/my-script.js`) and only considering absolute ones, such as `https://example.com/my-script.js`. It might make sense to include both for the `preload` resource hint.

Reply:
Updated queries to remove `WHERE domain IS NOT NULL`